### PR TITLE
[opentitantool] Add ultradebug and SPI flash support

### PR DESCRIPTION
**Pull request by cfrantz**
1. Integrate the ultradebug backend into opentitantool.
2. Create commands for manipulating GPIOs.
3. Create commands for manipulating SPI flash parts.

Signed-off-by: Chris Frantz <cfrantz@google.com>

### CODE REVIEW

File: sw/host/opentitanlib/src/spiflash/flash.rs [hunk 1]
```diff
@@ -208,7 +208,19 @@ impl SpiFlash {
...
+    }
+
+    /// Read into `buffer` from the SPI flash starting at `address`.
+    /// The `progress` callback will be invoked after each chunk of the read operation.
+    pub fn read_with_progress<F: FnMut(u32)>(
+        &self,
+        spi: &dyn Target,
+        mut address: u32,
+        buffer: &mut [u8],
+        mut progress: F,
```
Comment (mcy) [inline sw/host/opentitanlib/src/spiflash/flash.rs:222]: Mumble I would use `impl Fn(u32)` but up to you.

File: sw/host/opentitantool/Cargo.toml [hunk 2]
```diff
@@ -18,6 +18,7 @@ env_logger = "0.8.3"
 raw_tty = "0.1.0"
 regex = "1"
 nix = "0.17.0"
+progressing = "3.0.2"
```
Comment (mcy) [inline sw/host/opentitantool/Cargo.toml:21]: Not mandatory but https://docs.rs/indicatif/0.16.2/indicatif/ is much shinier IMO

File: sw/host/opentitantool/src/command/spi.rs [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+pub struct SpiSfdp {
+    #[structopt(
+        short,
+        long,
+        help = "Display raw SFDP bytes rather than the parsed struct."
+    )]
+    raw: Option<usize>,
+}
+
+fn hexdump(buf: &[u8]) {
```
Comment (mcy) [inline sw/host/opentitantool/src/command/spi.rs:32]: What does this function do? It's a little hard to tell so I"m not sure how to advise writing it in a simpler fashion.

File: sw/host/opentitantool/src/command/spi.rs [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+}
+
+impl CommandDispatch for SpiSfdp {
+    fn run(&self, transport: &mut dyn Transport) -> Result<Option<Box<dyn Serialize>>> {
+        transport.capabilities().request(Capability::SPI).ok()?;
+        let spi = transport.spi()?;
+
+        if let Some(length) = self.raw {
+            let mut buffer = Vec::new();
+            buffer.resize(length, 0);
```
Comment (mcy) [inline sw/host/opentitantool/src/command/spi.rs:64]: Looks like you want `vec![0; length]`?

Comment (cfrantz) [reply sw/host/opentitantool/src/command/spi.rs:64]: Done.

File: sw/host/opentitanlib/src/spiflash/flash.rs (hunk 1 above)
Comment (cfrantz) [reply sw/host/opentitanlib/src/spiflash/flash.rs:222]: Done.  (actually Fn(u32, u32), taking address and chunksize to make the progress updating functions easier).

File: sw/host/opentitantool/Cargo.toml (hunk 2 above)
Comment (cfrantz) [reply sw/host/opentitantool/Cargo.toml:21]: Indeed.  I changed to using that one.

File: sw/host/opentitantool/src/command/spi.rs (hunk 3 above)
Comment (cfrantz) [reply sw/host/opentitantool/src/command/spi.rs:38]: It prints a hexdump of a buffer in "exemplary form".  I added some comments explaining what's going on.

File: sw/host/opentitantool/src/command/spi.rs [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+                println!("  {}", String::from_utf8_lossy(&ascii));
+            }
+            // Print the offset.
+            print!("{:08x}:", i);
+        }
+        // Print each byte.
+        print!(" {:02x}", *v);
+        // For printable ASCII chars, place them in the ascii buffer.
+        // Replace all non-printable chars with a dot.
+        ascii[i % 16] = if *v >= 32 && *v < 127 { *v } else { b'.' };
```
Comment (mcy) [inline sw/host/opentitantool/src/command/spi.rs:53]: You can make the conditon be `*v == b' ' || v.is_ascii_graphic()`.

File: sw/host/opentitantool/src/command/spi.rs [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+//
+//  00000000: 53 46 44 50 06 01 02 ff 00 06 01 10 30 00 00 ff  SFDP........0...
+//  00000010: c2 00 01 04 10 01 00 ff 84 00 01 02 c0 00 00 ff  ................
+//  00000020: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
+//  00000030: e5 20 fb ff ff ff ff 3f 44 eb 08 6b 08 3b 04 bb  . .....?D..k.;..
+//
+// Note: This format can be consumed by `xxd -r` and converted back into binary.
+fn hexdump(buf: &[u8]) {
+    let mut ascii = [b'.'; 16];
+    for (i, v) in buf.iter().enumerate() {
```
Comment (mcy) [inline sw/host/opentitantool/src/command/spi.rs:40]: Maybe it makes sense to use `buf.chunks(16)` here, too? I hate to be like "nonono write this code" but I think you may find this bit of rust-foo none-the-less helpful:

```rust
for (i, word) in buf.chunks(16).enumerate() {
  let ascii = [b'.'; 16];
  print!("{:08x}:", i * 16);
  for (j, b) in word.iter().copied().enumerate() {
    if b == b' ' || b.is_ascii_graphic() { ascii[j] = b; }
    print!(" {:02x}", b);
  }
  for _ in 0..(16-i) {
    println!("   ");
  }
  println!("  {}", std::from_utf8(&ascii).unwrap());
}
```

Thoughts?

File: sw/host/opentitantool/src/command/spi.rs [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+        short = "n",
+        long,
+        default_value = "4096",
+        help = "Number of bytes to read."
+    )]
+    length: usize,
+    #[structopt(short, long, help = "Hexdump to stdout.")]
+    hexdump: bool,
+    #[structopt(name = "FILE", default_value = "-")]
+    filename: String,
```
Comment (mcy) [inline sw/host/opentitantool/src/command/spi.rs:144]: Should be PathBuf right?

File: sw/host/opentitantool/src/command/spi.rs [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+        flash.read_with_progress(&*spi, self.start, &mut buffer, |_, chunk| {
+            progress.inc(chunk as u64);
+        })?;
+        progress.finish();
+        let t1 = Instant::now();
+        let duration = (t1 - t0).as_secs_f64();
+        let bytes_per_sec = buffer.len() as f64 / duration;
+
+        if self.filename == "-" {
+            if self.hexdump {
```
Comment (mcy) [inline sw/host/opentitantool/src/command/spi.rs:172]: Do you want hexdump to work irrespective of which file is used?

File: sw/host/opentitantool/src/command/spi.rs [hunk 9]
```diff
@@ -0,0 +1,276 @@
...
+        let mut flash = SpiFlash::from_spi(&*spi)?;
+        flash.set_address_mode_auto(&*spi)?;
+
+        let progress = progress_bar(self.length as u64);
+        let t0 = Instant::now();
+        flash.erase_with_progress(&*spi, self.start, self.length, |_, chunk| {
+            progress.inc(chunk as u64);
+        })?;
+        progress.finish();
+        let t1 = Instant::now();
```
Comment (mcy) [inline sw/host/opentitantool/src/command/spi.rs:217]: You can write `t0.elapsed()`.

File: sw/host/opentitantool/src/command/spi.rs (hunk 6 above)
Comment (cfrantz) [reply sw/host/opentitantool/src/command/spi.rs:40]: Its shorter and I _do_ like that.  I'm somewhat ambivalent about the chunking -- I'm going to run through the entire array^W slice anyway.

The original version I wrote was a straight translation of what I have memorized for a C hexdump function.

File: sw/host/opentitantool/src/command/spi.rs (hunk 5 above)
Comment (cfrantz) [reply sw/host/opentitantool/src/command/spi.rs:53]: Sigh.  I know its "more correct" to use the helper functions, but the range of printable ASCII characters is fixed forever.  I suppose its more a statement about the timeframe in which I learned to code that I just happen to have memorized certain features of the ASCII character set.

File: sw/host/opentitantool/src/command/spi.rs (hunk 7 above)
Comment (cfrantz) [reply sw/host/opentitantool/src/command/spi.rs:144]: A rose by any other name...

File: sw/host/opentitantool/src/command/spi.rs (hunk 8 above)
Comment (cfrantz) [reply sw/host/opentitantool/src/command/spi.rs:172]: Originally, I'd only intended for hexdump to be valid with stdout, but there's really no reason for that restriction.  Fixed to hexdump either to stdout or to the chosen file.

File: sw/host/opentitantool/src/command/spi.rs (hunk 9 above)
Comment (cfrantz) [reply sw/host/opentitantool/src/command/spi.rs:217]: Done

**Review by mcy — APPROVED** [review]

**Pull request merged**